Invalid cloud written using laspy

## Description

When using the *lazrs* backend with *laspy* to write the attached point cloud, the resulting LAZ file is corrupted.

The attached point cloud has the following characteristics:

- LAS Specification: 1.4
- Point Data Record Format: 8
- Point count: 50 000

## Environment

- Python: 3.8
- OS: Windows

### Python dependencies

- *lazrs*: 0.6.0
- *laspy*: 2.5.4
- *laszip*: 0.2.3

## Steps to reproduce

Copy the following script in a file named `test_corrupted.py`:

```python
from io import BytesIO

import laspy
import pytest

CLOUD_PATH = "corruptible.laz"


@pytest.mark.parametrize(
    "backend",
    [laspy.LazBackend.Laszip, laspy.LazBackend.Lazrs, laspy.LazBackend.LazrsParallel],
)
def test_write_read_corruptible_correctly_loads_cloud(
    backend: laspy.LazBackend,
) -> None:
    cloud = laspy.read(CLOUD_PATH, laz_backend=backend)
    stream = BytesIO()
    with laspy.open(
        stream, mode="w", header=cloud.header, closefd=False, laz_backend=backend
    ) as writer:
        writer.write_points(cloud.points)
    stream.seek(0)
    laspy.read(stream, laz_backend=backend)
```

Download the LAZ file [corruptible.log](https://github.com/user-attachments/files/16284168/corruptible.log) in the same directory as the script and rename it to `corruptible.laz` (*GitHub* would not accept the .laz extension...)

Launch:

```console
pytest test_corrupted.py
```

## Expected result

The 3 tests pass.

## Actual result

The tests related to *lazrs* fail with the following error:

```
self = <laspy._compression.lazrsbackend.LazrsPointReader object at 0x0000019575A1AA90>, n = 50000

    def read_n_points(self, n: int) -> bytearray:
        point_bytes = bytearray(n * self.vlr.item_size())
>       self.decompressor.decompress_many(point_bytes)
E       lazrs.LazrsError: IoError: failed to fill whole buffer

..\Miniconda3\envs\psanp\lib\site-packages\laspy\_compression\lazrsbackend.py:95: LazrsError
```

--- a/src/encoders.rs
+++ b/src/encoders.rs
@@ -94,7 +94,7 @@ use crate::decoders::AC_MIN_LENGTH;
 use crate::models;
 use crate::models::DM_LENGTH_SHIFT;
 
-const AC_BUFFER_SIZE: usize = 1024;
+const AC_BUFFER_SIZE: usize = 4096;
 
 pub struct ArithmeticEncoder<T: Write> {
     out_buffer: Vec<u8>,

--- a/src/encoders.rs
+++ b/src/encoders.rs
@@ -87,17 +87,16 @@
 //                                                                           -
 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 
-use std::io::Write;
-
 use crate::decoders;
 use crate::decoders::AC_MIN_LENGTH;
 use crate::models;
 use crate::models::DM_LENGTH_SHIFT;
+use std::io::Write;
 
 const AC_BUFFER_SIZE: usize = 4096;
 
 pub struct ArithmeticEncoder<T: Write> {
-    out_buffer: Vec<u8>,
+    out_buffer: Box<[u8]>,
 
     out_byte: *mut u8,
     end_byte: *const u8,
@@ -110,10 +109,11 @@ pub struct ArithmeticEncoder<T: Write> {
 
 impl<T: Write> ArithmeticEncoder<T> {
     pub fn new(out_stream: T) -> Self {
-        let mut out_buffer = vec![0u8; 2 * AC_BUFFER_SIZE];
-        let out_byte = out_buffer.as_mut_ptr();
+        let mut out_buffer = vec![0u8; 2 * AC_BUFFER_SIZE].into_boxed_slice();
 
-        let end_byte = unsafe { out_buffer.as_ptr().add(2 * AC_BUFFER_SIZE) };
+        let out_byte = out_buffer.as_mut_ptr_range().start;
+        let end_byte = out_buffer.as_mut_ptr_range().end;
+
         Self {
             out_buffer,
             out_byte,
@@ -124,12 +124,17 @@ impl<T: Write> ArithmeticEncoder<T> {
         }
     }
 
+    #[inline]
+    fn end_of_buffer(&self) -> *const u8 {
+        self.out_buffer.as_ptr_range().end
+    }
+
     pub fn reset(&mut self) {
         self.base = 0;
         self.length = decoders::AC_MAX_LENGTH;
-        self.out_buffer = vec![0u8; 2 * AC_BUFFER_SIZE];
+        self.out_buffer.fill(0);
         self.out_byte = self.out_buffer.as_mut_ptr();
-        self.end_byte = unsafe { self.out_buffer.as_ptr().add(2 * AC_BUFFER_SIZE) };
+        self.end_byte = self.end_of_buffer();
     }
 
     pub fn done(&mut self) -> std::io::Result<()> {
@@ -155,18 +160,18 @@ impl<T: Write> ArithmeticEncoder<T> {
         }
         self.renorm_enc_interval()?;
 
-        let endbuffer = unsafe { self.out_buffer.as_mut_ptr().add(2 * AC_BUFFER_SIZE) };
-        if self.end_byte != endbuffer {
-            unsafe {
-                debug_assert!(
-                    (self.out_byte as *const u8) < self.out_buffer.as_ptr().add(AC_BUFFER_SIZE)
-                );
-                let slc: &[u8] = std::slice::from_raw_parts(
-                    self.out_buffer.as_ptr().add(AC_BUFFER_SIZE),
+        if self.end_byte != self.end_of_buffer() {
+            debug_assert!(
+                (self.out_byte.cast_const())
+                    < self.out_buffer.as_ptr().wrapping_add(AC_BUFFER_SIZE)
+            );
+            let slc = unsafe {
+                std::slice::from_raw_parts(
+                    self.out_buffer.as_ptr().wrapping_add(AC_BUFFER_SIZE),
                     AC_BUFFER_SIZE,
-                );
-                self.out_stream.write_all(&slc)?;
-            }
+                )
+            };
+            self.out_stream.write_all(&slc)?;
         }
 
         let buffer_size = self.out_byte as isize - self.out_buffer.as_ptr() as isize;
@@ -236,7 +241,6 @@ impl<T: Write> ArithmeticEncoder<T> {
             self.length >>= DM_LENGTH_SHIFT;
             x = model.distribution[sym as usize] * self.length;
             self.base = self.base.wrapping_add(x);
-            //self.base += x;
             self.length = model.distribution[(sym + 1) as usize] * self.length - x;
         }
 
@@ -361,69 +365,63 @@ impl<T: Write> ArithmeticEncoder<T> {
     }
 
     fn propagate_carry(&mut self) {
-        let endbuffer = unsafe { self.out_buffer.as_mut_ptr().add(2 * AC_BUFFER_SIZE) };
-        let mut b = if self.out_byte as *const u8 == self.out_buffer.as_ptr() {
-            unsafe { endbuffer.offset(-1) }
+        let mut p = if self.out_byte.cast_const() == self.out_buffer.as_ptr() {
+            self.end_of_buffer().wrapping_sub(1).cast_mut()
         } else {
-            unsafe { self.out_byte.offset(-1) }
+            self.out_byte.wrapping_sub(1)
         };
 
         unsafe {
-            while *b == 0xFFu8 {
-                *b = 0;
-                if b as *const u8 == self.out_buffer.as_ptr() {
-                    b = self
-                        .out_buffer
-                        .as_mut_ptr()
-                        .add(2 * AC_BUFFER_SIZE)
-                        .offset(-1);
+            while *p == 0xFFu8 {
+                *p = 0;
+                if p.cast_const() == self.out_buffer.as_ptr() {
+                    p = self.end_of_buffer().wrapping_sub(1).cast_mut()
                 } else {
-                    b = b.offset(-1);
+                    p = p.wrapping_sub(1);
                 }
-                debug_assert!(self.out_buffer.as_ptr() <= b);
-                debug_assert!(b < endbuffer);
-                debug_assert!(self.out_byte < endbuffer);
+                debug_assert!(self.out_buffer.as_ptr() <= p);
+                debug_assert!(p.cast_const() < self.end_of_buffer());
+                debug_assert!(self.out_byte.cast_const() < self.end_of_buffer());
             }
-            *b += 1;
+            *p += 1;
         }
     }
 
     fn renorm_enc_interval(&mut self) -> std::io::Result<()> {
-        let endbuffer = unsafe { self.out_buffer.as_mut_ptr().add(2 * AC_BUFFER_SIZE) };
         loop {
             debug_assert!(self.out_buffer.as_ptr() <= self.out_byte);
-            debug_assert!(self.out_byte < endbuffer);
-            debug_assert!((self.out_byte as *const u8) < self.end_byte);
+            debug_assert!(self.out_byte.cast_const() < self.end_of_buffer());
+            debug_assert!(self.out_byte.cast_const() < self.end_byte);
             unsafe {
                 *self.out_byte = (self.base >> 24) as u8;
-                self.out_byte = self.out_byte.offset(1);
+            }
+            self.out_byte = self.out_byte.wrapping_add(1);
 
-                if self.out_byte as *const u8 == self.end_byte {
-                    self.manage_out_buffer()?;
-                }
-                self.base <<= 8;
-                self.length <<= 8; // length multiplied by 256
-                if self.length >= AC_MIN_LENGTH {
-                    break;
-                }
+            if self.out_byte.cast_const() == self.end_byte {
+                self.manage_out_buffer()?;
+            }
+            self.base <<= 8;
+            self.length <<= 8; // length multiplied by 256
+            if self.length >= AC_MIN_LENGTH {
+                break;
             }
         }
         Ok(())
     }
 
     fn manage_out_buffer(&mut self) -> std::io::Result<()> {
-        let endbuffer = unsafe { self.out_buffer.as_mut_ptr().add(2 * AC_BUFFER_SIZE) };
-        if self.out_byte == endbuffer {
+        debug_assert!(self.out_byte.cast_const() == self.end_byte);
+
+        if self.out_byte.cast_const() == self.end_of_buffer() {
             self.out_byte = self.out_buffer.as_mut_ptr();
         }
-        unsafe {
-            let slc: &[u8] = std::slice::from_raw_parts(self.out_byte, AC_BUFFER_SIZE);
-            self.out_stream.write_all(&slc)?;
-            self.end_byte = self.out_byte.add(AC_BUFFER_SIZE);
 
-            debug_assert!(self.end_byte > self.out_byte);
-            debug_assert!(self.out_byte < endbuffer);
-        }
+        let slc = unsafe { std::slice::from_raw_parts(self.out_byte, AC_BUFFER_SIZE) };
+        self.out_stream.write_all(slc)?;
+        self.end_byte = self.out_byte.wrapping_add(AC_BUFFER_SIZE);
+
+        debug_assert!(self.end_byte > self.out_byte);
+        debug_assert!(self.out_byte.cast_const() < self.end_of_buffer());
         Ok(())
     }
 }

--- a/src/las/point6.rs
+++ b/src/las/point6.rs
@@ -411,6 +411,21 @@ pub mod v3 {
         gps_time: bool,
     }
 
+    impl Point6FieldFlags {
+        fn default_for_compression() -> Self {
+            Self {
+                z: false,
+                classification: false,
+                flags: false,
+                intensity: false,
+                scan_angle: false,
+                user_data: false,
+                point_source: false,
+                gps_time: false,
+            }
+        }
+    }
+
     impl Default for Point6FieldFlags {
         fn default() -> Self {
             Self {
@@ -1315,7 +1330,7 @@ pub mod v3 {
         fn default() -> Self {
             Self {
                 encoders: Point6Encoders::default(),
-                has_changed: Point6FieldFlags::default(),
+                has_changed: Point6FieldFlags::default_for_compression(),
                 current_context: 0,
                 contexts: [
                     Point6CompressionContext::default(),
@@ -1822,9 +1837,12 @@ pub mod v3 {
             if last_point.user_data != current_point.user_data {
                 self.has_changed.user_data = true;
             }
-
             let model = the_context.models.user_data[last_point.user_data as usize / 4]
-                .get_or_insert_with(|| ArithmeticModelBuilder::new(256).build());
+                .get_or_insert_with(|| {
+                    ArithmeticModelBuilder::new(256)
+                        .with_compression(true)
+                        .build()
+                });
             self.encoders
                 .user_data
                 .encode_symbol(model, u32::from(current_point.user_data))?;
@@ -1890,6 +1908,7 @@ pub mod v3 {
                 point_source: return_len_if_has_changed_else_0!(point_source),
                 gps_time: return_len_if_has_changed_else_0!(gps_time),
             };
+
             sizes.write_to(dst)?;
             Ok(())
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -120,12 +120,12 @@ pub use laszip::{
     LazItemRecordBuilder, LazItemType, LazVlr, LazVlrBuilder,
 };
 
-pub(crate) mod compressors;
-pub(crate) mod decoders;
-pub(crate) mod decompressors;
-pub(crate) mod encoders;
+pub mod compressors;
+pub mod decoders;
+pub mod decompressors;
+pub mod encoders;
 pub mod laszip;
-pub(crate) mod models;
+pub mod models;
 pub mod packers;
 pub mod record;
 

--- a/src/models.rs
+++ b/src/models.rs
@@ -233,7 +233,12 @@ impl<'a> ArithmeticModelBuilder<'a> {
         }
     }
 
-    pub fn build(self) -> ArithmeticModel {
+    pub fn with_compression(&mut self, state: bool) -> &'_ mut Self {
+        self.compress = state;
+        self
+    }
+
+    pub fn build(&self) -> ArithmeticModel {
         ArithmeticModel::new(self.symbols, self.compress, self.init_table)
     }
 }

